pkg/cmd/cli/cmd/version.go

```diff
@@ -0,0 +1,158 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	etcdversion "github.com/coreos/etcd/version"
+
+	"k8s.io/kubernetes/pkg/client/restclient"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	kubeversion "k8s.io/kubernetes/pkg/version"
+
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/version"
+
+	"github.com/spf13/cobra"
+)
+
+// f := clientcmd.New(pflag.NewFlagSet("", pflag.ContinueOnError))
+
+const (
+	versionLong = `
+Display client and server versions for Kubernetes and OpenShift.`
+)
+
+type VersionOptions struct {
+	BaseName         string
+	Config           clientcmdapi.Config
+	ClientConfig     *restclient.Config
+	OClient          *client.Client
+	KClient          *kclient.Client
+	Out              io.Writer
+	ClientCmdFactory *clientcmd.Factory
+
+	PrintEtcdVersion bool
+}
+
+// NewCmdVersion creates a command for displaying the version of this binary
+func NewCmdVersion(fullName string, f *clientcmd.Factory, out io.Writer, printEtcdVersion bool) *cobra.Command {
+	options := &VersionOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "version",
+		Short: "Display client and server versions",
+		Long:  versionLong,
+		Run: func(cmd *cobra.Command, args []string) {
+			options.BaseName = fullName
+			options.PrintEtcdVersion = printEtcdVersion
+
+			if err := options.Complete(f, args, out); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.RunVersion(); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func (o *VersionOptions) Complete(f *clientcmd.Factory, args []string, out io.Writer) error {
+	if len(args) > 0 {
+		return fmt.Errorf("no arguments should be passed")
+	}
+
+	o.Out = out
+	o.ClientCmdFactory = f
+
+	return nil
+}
+
+// RunVersion attempts to display client and server versions for Kubernetes and OpenShift
+func (o VersionOptions) RunVersion() error {
+	fmt.Fprintf(o.Out, "%s %v\n", o.BaseName, version.Get())
+	fmt.Fprintf(o.Out, "kubernetes %v\n", kubeversion.Get())
+	if o.PrintEtcdVersion {
+		fmt.Fprintf(o.Out, "etcd %v\n", etcdversion.Version)
+	}
+
+	if o.ClientCmdFactory == nil {
+		return nil
+	}
+
+	ocServerVersionChan := make(chan version.Info)
+	kubeServerVersionChan := make(chan kubeversion.Info)
+	versionErrorChan := make(chan error)
+	fatalErrorChan := make(chan error)
+	timeoutChan := make(chan bool, 1)
+
+	// max amount of time we want to wait for server to respond
+	var maxRequestDuration int32 = 10
+
+	// start goroutine to fetch openshift / kubernetes server version
+	go func(f *clientcmd.Factory) {
```

Comment:
This could be a function setting values and closing a channel to indicate, "done", right?  You wait for "done" then, inspect the values to limit the number of channels you have.